Improve publishing setup

gobbletsit opened the pull request:
This PR contains changes made to update and improve android library template example with removing the buildSrc groovy based setup and replacing it with env var based sonatype config setup.

At first I wanted to create a separate repo just for this type of setup (so you can decide which one to use for yourself) but decided against it so we can discuss the pros and cons of each approach. So far I haven't encountered a con with this one(used on JsonApiX) and I do obviously prefer it.

Also, I took the liberty to enrich the example with multi module based setup so it's easier for people to understand the publishing feature from a more "real world scenario" example.

**Note***
If we decide to go with this approach, then the handbook would need to be updated/tweaked as well.

JonatanPlesko commented on build.gradle:
```diff
@@ -39,6 +39,7 @@ subprojects {
 //    apply from: "$rootDir/detekt.gradle"
```
Commenting on a random line because somehow Bitbucket is better than Github, which doesn't let you comment on binary files.
.DS_Store shouldn't be committed I believe. Should probably add it to .gitignore.

gobbletsit replied:
Good catch 👍

JonatanPlesko commented on config.gradle:
```diff
@@ -0,0 +1,7 @@
+ext {
+    releaseConfig = [
+            "group"      : "com.infinum.libname",
```
I see we have a lot of placeholders that would technically pass even if not replaced. Maybe we need to "mark" them somehow to make it more obvious for the user of the template and make it stand out more as something which needs to be replaced.
Maybe something as simple as leaving a "// TODO" on every line where something needs to be checked or replaced?

gobbletsit replied:
Yeah, agreed, I should make it more obvious 👍

JonatanPlesko commented on libModule1/build.gradle:
```diff
@@ -0,0 +1,31 @@
+plugins {
```
Now that you added modules, wouldn't it make sense to also add a "sampleApp" sample module where you show how to include the library modules in the library's sample app? I think pretty much every library should have some kind of a sample app or module, so it makes sense to include a sample "sample" module in the template too.

gobbletsit replied:
I like the idea, will update on this as well 👍

JonatanPlesko commented on libModule2/build.gradle:
```diff
@@ -0,0 +1,31 @@
+plugins {
+    id "java-library"
+    id "kotlin"
```
As you're already showing a multi-module setup, how about having one Android and one Java module instead of two identical ones?
If you need some inspiration, you can take a look here:
https://github.com/infinum/android-complexify/tree/milestone/3.0.0

gobbletsit replied:
I can show that as well 👍

JonatanPlesko commented on libModule1/build.gradle:
```diff
@@ -0,0 +1,31 @@
+]
+
+java {
+    sourceCompatibility = JavaVersion.VERSION_1_7
+    targetCompatibility = JavaVersion.VERSION_1_7
```
Is there a reason for 1.7? Actually, is there even a reason for 1.8? 😄

gobbletsit replied:
Not sure how this happened, my bad, will change to 11 👍

JonatanPlesko commented on maven-publish.gradle:
```diff
@@ -0,0 +1,73 @@
+
+afterEvaluate {
+    publishing {
+        // gets specific props defined per each module which breaks the need for multiple(per module) maven-publish files
```
```suggestion
        // gets specific props defined per each module which breaks the need for multiple (per module) maven-publish files
```

JonatanPlesko commented on maven-publish.gradle:
```diff
@@ -0,0 +1,73 @@
+                version = properties.version
+                artifactId = properties.artifactId
+
+//                artifact bundleReleaseAar
```
Maybe a comment about when to uncomment this?

gobbletsit replied:
Probably makes sense even though it's described in handbook. That's what I meant with the note in description about changing the handbook chapter as well.

antunflas replied:
Since we include this file in all projects, we can't really comment/uncomment things to make it work if project have both Android and non-Android modules.
Let's either check if project has plugin with id `com.android.library`/`com.android.application` or let's introduce two separate files. I prefer if check, do you see any problem with it?

gobbletsit replied:
@antunflas No I like the idea 👍

JonatanPlesko commented on maven-publish.gradle:
```diff
@@ -0,0 +1,73 @@
+                    developers {
+                        developer {
+                            id = 'malicombo'
+                            name = 'Mali Čombo'
```
What is even Čombo 😂

gobbletsit replied:
A Čovjek with a big head :D

antunflas replied:
@JonatanPlesko @schalke95 Let's also discuss if we're going to have one company wide developer (android@infinum.com) on all the libraries or everyone would update this list whenever they made a change.
Afaik this isn't connected to any account, just for informational purposes.

JonatanPlesko commented on maven.gradle:
```diff
@@ -1,12 +1,27 @@
-    instance.load()
-    instance
+def properties = new Properties()
+def propsFile = file("$rootDir/publish.properties")
```
We don't have any examples for this file, what should be in it? 🤔

gobbletsit replied:
It's described in the handbook, that's why I saw no point in explaining here as its not just a coupe of lines.

JonatanPlesko replied:
TBH I'm not against putting all that info here as well, it's easier to delete 20 lines of docs than it is to open a handbook and search for it. Though I also agree it is something that the handbook section on library publishing should contain.

JonatanPlesko commented on maven.gradle:
```diff
@@ -1,12 +1,27 @@
+* /gradlew deploy -PsonatypeUsername={yourMavenUsername} -PsonatypePassword={yourMavenPassword}.
+* 2. properties["sonatype.user"]: use property defined in publish.properties file in the root directory. If you have that file in the root
+* of the project you can run -> ./gradlew deploy
+* 3. System.getenv("SONATYPE_PASS"): use global variable defined in the system. No need to create publish.properties file, you can simply
```
Maybe it would be better to generically explain what are the ways in which these can be provided, this is a bit inconsistent, here you mention the "pass" property but before you were talking about "user".

gobbletsit replied:
I can do that 👍

gobbletsit replied:
Aha, these were only examples but considering it wasn't clear I should update it 👍

JonatanPlesko reviewed (commented):
I like this, good improvements, and thank you for the proactive change! Let's discuss this further, and I'd also like one of the other guys to give their feedback on this.

gobbletsit commented:
@antunflas @schalke95 I would like to start with implementing @JonatanPlesko suggestions so would like to check if you're gonna have a look before I do so?

antunflas commented on config.gradle:
```diff
@@ -0,0 +1,7 @@
+    releaseConfig = [
+            "group"      : "com.infinum.libname",
+            "version"    : "0.0.1",
+            "versionCode": 0 * 10000 + 0 * 100 + 1
```
Do we need version code in the libraries?

gobbletsit replied:
Sry, I've missed this one. Not sure if we need it. I know it's present on the libraries I've encountered so I suppose it is indeed needed.

antunflas replied:
I'm pretty sure it isn't needed, and the most popular Android libraries such as Timber, Room or Glide don't have it.
What are the examples of libraries that have version code defined?

gobbletsit replied:
Then I'll eliminate it :) . Sentinel, JsonApiX, Compexify.

antunflas reviewed (commented):
👍 I agree with all @JonatanPlesko comments

gobbletsit commented:
@antunflas @JonatanPlesko I've also added a sample publish.properties file and added a comment explaining the need based on Option 3 described in maven.gradle file. Also it's ignored so you will not be able to see it here.

JonatanPlesko commented on libModule2/proguard-rules.pro:
```diff
@@ -0,0 +1,21 @@
+# Add project specific ProGuard rules here.
```
Is there a particular reason to have the proguard rules file?

gobbletsit replied:
No particular reason, can be removed if we wish to.

simicmarin replied:
I'm fine with leaving this since android library modules usually have it.

JonatanPlesko commented:
> @antunflas @JonatanPlesko I've also added a sample publish.properties file and added a comment explaining the need based on Option 3 described in maven.gradle file. Also it's ignored so you will not be able to see it here.

If it's ignored, where exactly are we supposed to see it? 🤔

JonatanPlesko commented on maven.gradle:
```diff
@@ -1,27 +1,44 @@
-* 3. System.getenv("SONATYPE_PASS"): use global variable defined in the system. No need to create publish.properties file, you can simply
-* run -> ./gradlew deploy.
-*/
+* There are 3 possible approaches to retrieving previously defined sonatype credentials and setting it on a sonatype file:
+*
+*    Option 1:
+*       - findProperty("SOME_PROPERTY") - use parameter from the command (if attached) so you can run ->
+*         ./gradlew deploy -PsonatypeSOME_PROPERTY_NAME={SOME_PROPERTY_VALUE} -PsonatypeSOME_PROPERTY_NAME_2={SOME_PROPERTY_VALUE_2}
+*
+*    Option 2:
+*       - properties["SOME_PROPERTY"]: use property defined in publish.properties file in the root directory(if going with this option the
+*         file must first be manually created as described in the handbook) so you can run -> ./gradlew deploy
+*
+*    Option 3:
+*       - System.getenv("SONATYPE_PASS"): use global variable defined in the system. With this approach there is no need to create a
+*         publish.properties file so you can simply run -> ./gradlew deploy
+* */
```
I do wonder if we should even present so many options or if we should just decide on one as the recommended one and go for it 🤔 I'm pretty sure that option 2 is the simplest one.

gobbletsit replied:
Most probably the simplest one but wanted to provide the users with options since I did use Option 3 for JsonApiX. This was actually the motivation behind the current improvement but we can probably discuss it more.

JonatanPlesko commented on sample/build.gradle:
```diff
@@ -0,0 +1,53 @@
+dependencies {
+
+    // TODO - replace with your module names accordingly
+    // Comment this out for testing local deploy
+    implementation project(":libModule1")
+    implementation project(":libModule2")
+
+    // TODO - replace with your module packages accordingly
+    // Uncomment this for testing local deploy
+//    implementation packages.android-libname.libModule1
+//    implementation packages.android-libname.libModule2
```
This is quite useful 👍

JonatanPlesko commented on sample/build.gradle:
```diff
@@ -0,0 +1,53 @@
+//    implementation packages.android-libname.libModule1
+//    implementation packages.android-libname.libModule2
+
+    implementation 'com.android.support:appcompat-v7:28.0.0'
+    implementation 'com.android.support.constraint:constraint-layout:2.0.4'
```
Can we use AndroidX instead? 😂

gobbletsit replied:
Sure :D

simicmarin replied:
Done as a part of https://github.com/infinum/android-library-template/pull/3 PR

JonatanPlesko reviewed (approved):
Approved, but I believe this should be merged together with #3 when that one is ready.

simicmarin commented:
> Approved, but I believe this should be merged together with #3 when that one is ready.

@JonatanPlesko #3 Is merged, I would say this one is now ready to be merged into main.

JonatanPlesko commented:
Sure 👍
